### PR TITLE
Add new backend type for Intel heterogeneous computation platform.

### PR DESCRIPTION
**Pull request by chengjunlu**
Add a new device type 'XPU' ('xpu' for lower case) to PyTorch. Changes are needed for code related to device model and kernel dispatch, e.g. DeviceType, Backend and DispatchKey etc.

#48246

### CODE REVIEW

**Comment by gujinghui** [comment]
@ezyang @VitalyFedyunin
Could you help review this PR for XPU and give comments? Thanks a lot.

@Jianhui-Li @jgong5

File: aten/src/ATen/TensorIterator.cpp [hunk 1]
```diff
@@ -376,10 +376,8 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
     // Checks all tensors are on the same device, if requested
     if (config.check_all_same_device_) {
       // Handles CPU scalars on CUDA kernels that support them
-      if (common_device.is_cuda() &&
-          config.allow_cpu_scalars_ &&
-          !op.is_output &&
-          op.tensor.dim() == 0 &&
+      if ((common_device.is_cuda() || common_device.is_xpu()) &&
```
Comment (ezyang) [inline aten/src/ATen/TensorIterator.cpp:379]: This change looks a little questionable; are you using TensorIterator to implement some XPU kernels?

File: c10/core/Backend.h [hunk 2]
```diff
@@ -31,15 +31,18 @@ enum class Backend {
...
   SparseCUDA,
   SparseHIP,
+  SparseXPU,
   MSNPU,
   XLA,
   Vulkan,
   Metal,
   QuantizedCPU,
   QuantizedCUDA,
+  QuantizedXPU,
```
Comment (ezyang) [inline c10/core/Backend.h:45]: Is QuantizedXPU actually going to be a thing?

File: torch/csrc/jit/serialization/pickler.cpp [hunk 3]
```diff
@@ -603,7 +603,9 @@ WriteableTensorData getWriteableTensorData(
   result.tensor_ = tensor;
   result.size_ = tensor.storage().nbytes();
   // TODO HIP support
-  if (tensor.storage().device_type() == DeviceType::CUDA && to_cpu) {
+  if ((tensor.storage().device_type() == DeviceType::CUDA ||
+       tensor.storage().device_type() == DeviceType::XPU) &&
+      to_cpu) {
```
Comment (ezyang) [inline torch/csrc/jit/serialization/pickler.cpp:608]: This looks highly questionable. Maybe the correct fix is `tensor.storage().device_type() != DeviceType::CPU`

**Comment by ezyang** [comment]
The PR looks fine. I'm mostly interested in whether or not there is some evidence (it doesn't have to be strong evidence, I just don't want it to be cargo culting) whether or not some of the functionality introduced by this PR will be used. In particular:

* Sparse XPU
* Quantized XPU
* AutogradXPU key (you only need this if you intend to override backward formulas of operators ONLY for XPU)
* TorchScript support for testing if a tensor is XPU

File: aten/src/ATen/TensorIterator.cpp (hunk 1 above)
Comment (gujinghui) [reply aten/src/ATen/TensorIterator.cpp:379]: Yes. We are using TensorIterator in some XPU kernels.

File: torch/csrc/jit/serialization/pickler.cpp (hunk 3 above)
Comment (gujinghui) [reply torch/csrc/jit/serialization/pickler.cpp:608]: We are not sure whether other devices, like HIP, XLA, already implemented all below tensor and storage APIs. Could you give more detailed suggestion here?

File: c10/core/Backend.h (hunk 2 above)
Comment (gujinghui) [reply c10/core/Backend.h:45]: Yes. We are already using QuantizedXPU to implement the low precise in XPU.

File: torch/csrc/jit/serialization/pickler.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/jit/serialization/pickler.cpp:608]: They're indirectly supported via the call to `.cpu()`, which these support.

Comment (gujinghui) [reply torch/csrc/jit/serialization/pickler.cpp:608]: Thanks for quick reply. We will follow your suggestion soon. Thanks.

**Comment by gujinghui** [comment]
> The PR looks fine. I'm mostly interested in whether or not there is some evidence (it doesn't have to be strong evidence, I just don't want it to be cargo culting) whether or not some of the functionality introduced by this PR will be used. In particular:
> 
> * Sparse XPU
> * Quantized XPU
> * AutogradXPU key (you only need this if you intend to override backward formulas of operators ONLY for XPU)
> * TorchScript support for testing if a tensor is XPU

@ezyang 
All of above will be supported in XPU.
For Sparse XPU, we already have plan and WIP.
For Quantized XPU, we already implemented and run several models in low precise successfully.
For AutogradXPU, yes, we will override several backward formulas only for XPU.
For TorchScript testing support, yes, we are porting all PyTorch test suites to ensure same level quality of our extension, in which we needs to check the tensor attribute.

File: torch/csrc/jit/serialization/pickler.cpp (hunk 3 above)
Comment (chengjunlu) [reply torch/csrc/jit/serialization/pickler.cpp]: The code has been changed to `tensor.storage().device_type() != DeviceType::CPU` for more general cases

File: aten/src/ATen/TensorIterator.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/TensorIterator.cpp:379]: OK. FYI there's likely going to be some big refactors to TensorIterator (changing UI) coming this half. cc @heitorschueroff

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/TensorIterator.cpp (hunk 1 above)
Comment (gujinghui) [reply aten/src/ATen/TensorIterator.cpp:379]: Thanks for the info.

@chengjunlu @arthuryuan1987

**Pull request closed without merging**